### PR TITLE
Create PULL_REQUEST_TEMPLATE.md [skip ci]

### PR DESCRIPTION
**Pull request by 183amir**
@conda-forge/core I create a pull request template. What do you think?

### CODE REVIEW

**Comment by 183amir** [comment]
@ocefpaf I opened a new one here.

**Comment by ocefpaf** [comment]
> @ocefpaf I opened a new one here.

No problems. Please see me comments on #549, they are minor easy to fix.

@183amir I believe we can put the template files (PR and Issues) in a `.github` directory to avoid the clutter.

**Comment by jakirkham** [comment]
Nice idea. Will try to review later.

I believe one can use `CONTRIBUTING.md` and GitHub will link it to the top of every issue and PR.

**Comment by 183amir** [comment]
True. There is also `ISSUE_TEMPLATE.md`

**Comment by ocefpaf** [comment]
I would not worry about `CONTRIBUTING.md` and `ISSUE_TEMPLATE.md` right now. However, this PR template you put together is **very** useful and long overdue! Let's get that on the repo ASAP.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+Recipes added with this pull request (max 5):
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: Not sure that we should have a hard limit on this. For some people, dumping everything in one go is the easiest way to get started. Admittedly, we may end up breaking it up in the end, but it would be better for reviewers and contributors to work together to strike a healthy balance for this on a case-by-case basis.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ 
+ - package1
+ - package2
+ 
+I have checked the following in my recipes:
+* [ ] My files end with only 1 newline character, no more, no less.
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: 👍

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+* [ ] My recipe follow the same style as [example](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml) as much as possible.
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: We might need to clean that up to match the style we want first. It fits an older acceptable style. Also, we have been discussing having a few, which we should work on somewhere.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+* [ ] The license field also specifies the number of the license if applicable (e.g. `GPLv3` instead of `GPL` or `BSD 3-clause` instead of `BSD`).
+* [ ] My recipe has tests.
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: Should we be a bit more specific about what acceptable tests are? For instance, running `make check`, `ctest`, what have you. Also, verifying all files were installed properly. Possibly, a simple program to exercise the package (especially if there is no `make check`, etc.).

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+* [ ] I have looked at [pinned packages](https://github.com/conda-forge/staged-recipes/wiki/Pinned-dependencies) and pinned those packages as stated.
+* [ ] The `summary` just explains the package and does not include the package name. (e.g. Instead of saying `Jupyterhub is a multi-user server for Jupyter notebooks` just say `Multi-user server for Jupyter notebooks`)
+
+If you build for Windows too:
+* [ ] I have read [VC features](https://github.com/conda-forge/staged-recipes/wiki/VC-features)
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: ...would add " and implemented".

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+* [ ] The license field also specifies the number of the license if applicable (e.g. `GPLv3` instead of `GPL` or `BSD 3-clause` instead of `BSD`).
+* [ ] My recipe has tests.
+* [ ] I have looked at [pinned packages](https://github.com/conda-forge/staged-recipes/wiki/Pinned-dependencies) and pinned those packages as stated.
+* [ ] The `summary` just explains the package and does not include the package name. (e.g. Instead of saying `Jupyterhub is a multi-user server for Jupyter notebooks` just say `Multi-user server for Jupyter notebooks`)
+
+If you build for Windows too:
+* [ ] I have read [VC features](https://github.com/conda-forge/staged-recipes/wiki/VC-features) 
+
+If recipe uses make or cmake or ctest:
+* [ ] I have also added `make check` or similar if applicable.
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: Sorry hadn't gotten here yet. This looks good.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+If recipe builds a library:
+* [ ] I have enabled both static and shared libraries.
+
+If recipe builds some C/C++, Fortran or OMP code:
+* [ ] I have not included `gcc` or `libgcc` in `requirements`. Exceptions can be made but must be tested first with `gcc`/`clang` that is already installed in our CI machines.
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: Would rephrase. Currently Fortran and OMP are ok uses of the packages. So, we should make an exception. Though this is something we strive to eliminate.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+If it is a Python PyPI package:
+* [ ] I have used `pip` to install it and my recipe has these elements:
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: Yeah, we really need a decision on this issue ( https://github.com/conda-forge/staged-recipes/issues/528 ) first.

Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: Though I am in favor that being a nice, simple solution forward.

Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: 1. `python setup.py install --single-version-externally-managed --record=record.txt`
2. `python setup.py install`

For now, let's document what we do, which is have people try 1, but if they don't use `setuptools` or it fails fallback to 2.

**Comment by jakirkham** [comment]
Took a look. Largely in favor of most of it. Some things need tweaks. Here are a couple big points.

* Python install: we should stick with the status quo until we make a decision on this ( https://github.com/conda-forge/staged-recipes/issues/528 ).
* Compilers: same story (though admittedly that changed very recently).
* Example recipe: should update to be a better representation of what we do now.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
+Recipes added with this pull request (max 5):
+ 
+ - package1
+ - package2
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: Not clear on what this listing is accomplishing.

**Comment by jakirkham** [comment]
Took a look. Largely in favor of most of it. Some things need tweaks. Here are a couple big points.

* Python install: we should stick with the status quo until we make a decision on this ( https://github.com/conda-forge/staged-recipes/issues/528 ).
* Compilers: same story (though admittedly that changed very recently).
* Example recipe: should update to be a better representation of what we do now.

**Comment by jakirkham** [comment]
> I believe we can put the template files (PR and Issues) in a .github directory to avoid the clutter.

👍  Correct. See [ref]( https://github.com/blog/2111-issue-and-pull-request-templates ).

**Comment by 183amir** [comment]
I will be out till Monday. Please feel free to open another pull request based on this one and add your comments directly there if you want to merge this as soon as possible.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 9 above)
Comment (183amir) [inline .github/PULL_REQUEST_TEMPLATE.md]: It helps to see which packages are being added without having to go through the "files changed" tab.
P.S. I pushed a new commit taking your comments in consideration.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    - python
+```
+* [ ] Test commands such as `nosetests package` or `py.test` are added:
+```yaml
+test:
+  requires:
+  - nose
+
+  commands:
+  - nosetests -sv package
```
Comment (jakirkham) [inline .github/PULL_REQUEST_TEMPLATE.md]: IMHO the point of testing is not really to run CI for the packages that are given to us. Instead it is here to guarantee the following things.

1. The compiler didn't mess something up.
2. Something didn't go wrong in the linking stage.
3. That the package will work in the environment defined.

The first two don't really apply to pure Python packages so let's ignore that for now. As far as solving the last one, importing modules seems to be sufficient for the most part. Running the full test suite is something the developer(s) hopefully are doing on releases at least if not during development as part of CI. Thus, there isn't really much value in us doing it to for pure Python packages. If it does have compiled portions, that is a totally different story.

In the end, there may be a few packages that need something like `run_test.py` to do a few more things, but that will probably be determined after it is a feedstock.

In short, I don't think this should be a hard requirement for pure Python packages.

**Comment by 183amir** [comment]
@jakirkham since you have written the guidelines and this is really urgent for people adding adding new recipes, why don't you create another pull request based on this one and add your comments directly to it? so that we can have this merged as fast as possible. We can always iterate this later.

**Comment by ocefpaf** [comment]
Closing as "stale."
